lib/__flake-version-info.nix

```diff
@@ -0,0 +1,17 @@
+# lib overlay to be used by the nixpkgs & nixpkgs/lib flakes
+# to provide meaningful values for `lib.trivial.version` et al..
+#
+# Internal and subject to change, don't use this anywhere else!
+
+self: # from the flake
+
+finalLib: prevLib: # lib overlay
+
+{
+  trivial = prevLib.trivial // {
+    versionSuffix =
+      ".${finalLib.substring 0 8 (self.lastModifiedDate or self.lastModified or "19700101")}.${self.shortRev or "dirty"}";
```

Comment:
This feels very arbitrary, but I guess it was done that way before. This should be documented as using an unspecified format.

Reply:
Hyrum's law:

> With a sufficient number of users of an API,
> it does not matter what you promise in the contract:
> all observable behaviors of your system
> will be depended on by somebody.

I figure for something like this, we might as well promise not to change it, and even document the exact format.

Reply:
> This feels very arbitrary, but I guess it was done that way before

@infinisil Correct, this is precisely the format everybody had when using `nixpkgs.lib.nixosSystem`.

It's not entirely unspecified or arbitrary though. This is the format you also get when using a channel tarball to build your system:

```
$ pwd                               
/nix/var/nix/profiles/per-user/root/channels/nixos
$ cat .version-suffix 
.20231011.97d028b
$ nix-instantiate --eval -A lib.trivial.versionSuffix 
".20231011.97d028b"
```

Only when importing the nixpkgs flake as path rather than as git input, you'll get something else, i.e. `YYYYMMDD.dirty`. I'd argue that `YYYYMMDD.dirty` is still better than `pre-git` though since you get a little more information at least (what we can't implement with lib.trivial and channel tarballs I think). However we can switch back to `pre-git` here if you prefer that (while I prefer the former option, I don't have a strong opinion though because you should only encounter that when testing something with a local checkout).

I'll update the documentation accordingly as soon as we've agreed on everything else.

> we might as well promise not to change it

@roberth fwiw Hyrum's law should already apply already since the format was/is used by `nixpkgs.lib.nixosSystem`.

Reply:
I think it's fine to provide as much info as possible, let's treat the version suffix as an arbitrary string without guarantees.

Reply:
Where are you thinking of documenting this though? Because it wouldn't really fit into the docs for `lib.trivial.versionSuffix`, since it would be specific to Flakes. I guess we could close an eye for that, but I feel like anything documented in `lib` should apply generally.

Reply:
> Where are you thinking of documenting this though?

Didn't bother to find a suitable place yet because I wanted the other stuff to get resolved first :upside_down_face: 

> Because it wouldn't really fit into the docs for lib.trivial.versionSuffix, since it would be specific to Flakes

Oh so we have some basic documentation about that already? I'm afraid I fail to find it right now, but may I ask where?

> since it would be specific to Flakes

It's even just the "edge" case of a local path override where the suffix looks different, so I'd argue that it would be OKish.

> but I feel like anything documented in lib should apply generally.

I mean, we effectively have two ways of using lib (this change also applies to the `nixpkgs/lib` flake itself), so it seems OK to me to mention a slightly different behavior when using one of them.

Reply:
> Oh so we have some basic documentation about that already? I'm afraid I fail to find it right now, but may I ask where?

[We do indeed](https://nixos.org/manual/nixpkgs/unstable/#function-library-lib.trivial.versionSuffix) :slightly_smiling_face: ([source](https://github.com/NixOS/nixpkgs/blob/8ad872b2f923793abc8822cce462fbad03dec237/lib/trivial.nix#L200))

Reply:
That already doesn't specify what the suffix contains, right? And this change doesn't remove that documentation, so I think it's fine as is?

Reply:
Good point. This string is for informational purposes only (i.e. when I want to check what's deployed on my system, then this is helpful), so perhaps it's not a good idea to start documenting a rather arbitrary format in there now. Hence, resolving this thread now.